[WIP] Parallelizing Histogram Creation

Bouncner opened the pull request:
This PR is a first step to better parallelize the generation of histograms.
It's only a "first step" as we plan to improve merging as well as string handling. But I'd like to keep this PR small.

The current status roughly makes the TPC-H generation twice as fast (no matter how many cores).

Bouncner committed:
Initial

Bouncner committed:
Small additions

Bouncner committed:
Centralize bin count

Bouncner committed:
Lint and format

fschmllr commented on src/lib/statistics/statistics_objects/equal_distinct_count_histogram.cpp:
```diff
@@ -20,60 +25,161 @@
+  segments_to_process.reserve(std::min(chunk_count, MIN_CHUNK_COUNT_TO_INCLUDE));
+
+  // In average, we sample every 2.5th segment. Previous analyses of Hyrise histograms showed that the accuracy of
+  // sample histograms quickly deteriogates with sampling rates below 33 %. We thus choose a safer rate of ~40%.
```
```suggestion
  // sample histograms quickly deteriorates with sampling rates below 33 %. We thus choose a safer rate of ~40%.
```

fschmllr commented on src/lib/statistics/table_statistics.cpp:
```diff
@@ -25,12 +25,7 @@ namespace hyrise {
-   * (for tables with >= 200m rows) are created.
-   */
-  const auto histogram_bin_count = std::min<size_t>(100, std::max<size_t>(5, table.row_count() / 2'000));
+  const auto bin_count = EqualDistinctCountHistogram<ColumnDataType>::determine_bin_count(lineitem_table->row_count());
```
```suggestion
  const auto bin_count = EqualDistinctCountHistogram<ColumnDataType>::determine_bin_count(table->row_count());
```

Bouncner committed:
Apply suggestions from code review

Co-authored-by: Florian Schmeller <26191888+fschmllr@users.noreply.github.com>

Bouncner committed:
Update src/lib/statistics/table_statistics.cpp

Co-authored-by: Florian Schmeller <26191888+fschmllr@users.noreply.github.com>

Bouncner committed:
Update tpch_data_micro_benchmark.cpp

Fix column type.

Bouncner committed:
Update table_statistics.cpp

Fix column type in table statistics.

Bouncner committed:
Format

Bouncner committed:
Update equal_distinct_count_histogram.cpp

fschmllr commented on src/lib/statistics/statistics_objects/equal_distinct_count_histogram.cpp:
```diff
@@ -20,60 +25,166 @@
+  auto result = ValueDistributionVector<T>{};
+  if (chunk_count > 0) {
+    // We determine the recursion steps (i.e., merge levels) that we want to parallelize. We try to create up to 2x the
+    // number workers to fully utilize a system with a bid of straggler mitigation (thus 2x) while not overloading the
```
```suggestion
    // number of workers to fully utilize a system with a bid of straggler mitigation (thus 2x) while not overloading the
```

fschmllr replied:
```suggestion
    // number workers to fully utilize a system with a bit of straggler mitigation (thus 2x) while not overloading the
```

Bouncner committed:
Apply suggestions from code review

Co-authored-by: Florian Schmeller <26191888+fschmllr@users.noreply.github.com>

Bouncner committed:
Use worker_count instead of topology

Bouncner committed:
Include node queue scheduler header and fix lint issue

Bouncner committed:
Use -> over . to access node queue scheduler method

Bouncner committed:
Merge, merge, merge

Bouncner committed:
Merge branch 'master' into martin/perf/parallel_histograms_simple

Bouncner committed:
Simplify code

Bouncner committed:
Format

Bouncner committed:
Fix

Bouncner committed:
Fix issue

Bouncner commented:
## Histogram Creation on Nemea (56 threads)

### Master
```
- Adding tables to StorageManager and generating table statistics
-  Added 'region' (2 ms 633 µs)
-  Added 'nation' (4 ms 777 µs)
-  Added 'supplier' (171 ms 776 µs)
-  Added 'customer' (2 s 850 ms)
-  Added 'part' (2 s 942 ms)
-  Added 'partsupp' (17 s 225 ms)
-  Added 'orders' (25 s 591 ms)
-  Added 'lineitem' (1 min 8 s)
- Adding tables to StorageManager and generating table statistics done (1 min 8 s)
```

### This PR
```
- Adding tables to StorageManager and generating table statistics
-  Added 'region' (387 µs 675 ns)
-  Added 'nation' (2 ms 270 µs)
-  Added 'supplier' (8 s 255 ms)
-  Added 'partsupp' (13 s 42 ms)
-  Added 'orders' (22 s 78 ms)
-  Added 'lineitem' (22 s 323 ms)
-  Added 'customer' (22 s 327 ms)
-  Added 'part' (22 s 708 ms)
- Adding tables to StorageManager and generating table statistics done (22 s 708 ms)
```

Bouncner committed:
Making clang-tidy at least a bit happier

Bouncner committed:
auto for the win

Bouncner committed:
Update equal_distinct_count_histogram.cpp

dey4ss commented on src/lib/statistics/statistics_objects/equal_distinct_count_histogram.cpp:
```diff
@@ -27,53 +33,169 @@ using namespace hyrise;
-  return value_distribution;
+  for (auto chunk_id = ChunkID{0}; chunk_id < chunk_count; ++chunk_id) {
+    const auto chunk = table.get_chunk(chunk_id);
+    Assert(chunk, "Physically deleted chunk should not reach this point, see get_chunk / #1686.");
```
That's not true. This code will be run on base tables, not on output tables of a `GetTable` operator.

Bouncner replied:
But chunks can still be deleted?

Bouncner committed:
Small changes

Bouncner committed:
Use std::move when merging to avoid string copies.

Bouncner committed:
Remove std::move. Suprisingly slower
